overlord/hookstate/ctlcmd/get.go

```diff
@@ -121,3 +129,178 @@ func (c *getCommand) Execute(args []string) error {
+		if err == nil {
+			return value, true, nil
+		}
+		if config.IsNoOption(err) {
+			if !c.Typed {
+				value = ""
+			}
+			return value, false, nil
+		}
+		return value, false, err
+	})
+}
+
+type ifaceHookType int
+
+const (
+	preparePlugHook ifaceHookType = iota
+	prepareSlotHook
+	connectPlugHook
+	connectSlotHook
+	unknownHook
+)
+
+func interfaceHookType(hookName string) (ifaceHookType, error) {
+	if strings.HasPrefix(hookName, "prepare-plug-") {
+		return preparePlugHook, nil
+	} else if strings.HasPrefix(hookName, "connect-plug-") {
+		return connectPlugHook, nil
+	} else if strings.HasPrefix(hookName, "prepare-slot-") {
+		return prepareSlotHook, nil
+	} else if strings.HasPrefix(hookName, "connect-slot-") {
+		return connectSlotHook, nil
+	}
+	return unknownHook, fmt.Errorf("unknown hook type")
+}
+
+func validatePlugOrSlot(attrsTask *state.Task, plugSide bool, plugOrSlot string) error {
+	// check if the requested plug or slot is correct for given hook.
+	attrsTask.State().Lock()
+	defer attrsTask.State().Unlock()
+
+	var name string
+	var err error
+	if plugSide {
+		var plugRef interfaces.PlugRef
+		if err = attrsTask.Get("plug", &plugRef); err == nil {
+			name = plugRef.Name
+		}
+	} else {
+		var slotRef interfaces.SlotRef
+		if err = attrsTask.Get("slot", &slotRef); err == nil {
+			name = slotRef.Name
+		}
+	}
+	if err != nil {
+		return fmt.Errorf(i18n.G("failed to find plug/slot data in the attrs task"))
+	}
+	if name != plugOrSlot {
+		return fmt.Errorf(i18n.G("unknown plug/slot %s"), plugOrSlot)
+	}
+	return nil
+}
+
+func attributesTask(context *hookstate.Context) (*state.Task, error) {
+	var attrsTaskID string
+	context.Lock()
+	defer context.Unlock()
+
+	if err := context.Get("attrs-task", &attrsTaskID); err != nil {
+		return nil, err
+	}
+
+	st := context.State()
+
+	attrsTask := st.Task(attrsTaskID)
+	if attrsTask == nil {
+		return nil, fmt.Errorf(i18n.G("failed to find attrs task"))
+	}
+
+	return attrsTask, nil
+}
+
+func (c *getCommand) getInterfaceSetting(context *hookstate.Context, plugOrSlot string) error {
+	// Make sure get :<plug|slot> is only supported during the execution of interface hooks
+	hookType, err := interfaceHookType(context.HookName())
+	if err != nil {
+		return fmt.Errorf(i18n.G("interface attributes can only be read during the execution of interface hooks"))
+	}
+
+	var attrsTask *state.Task
+	attrsTask, err = attributesTask(context)
+	if err != nil {
+		return fmt.Errorf(i18n.G("failed to find attrs task: %q"), err)
+	}
+
+	if c.ForcePlugSide && c.ForceSlotSide {
+		return fmt.Errorf("cannot use --plug and --slot together")
+	}
+
+	isPlugSide := (hookType == preparePlugHook || hookType == connectPlugHook)
+	isSlotSide := (hookType == prepareSlotHook || hookType == connectSlotHook)
+	if err = validatePlugOrSlot(attrsTask, isPlugSide, plugOrSlot); err != nil {
+		return err
+	}
+
+	var which string
+	if (isPlugSide && !c.ForceSlotSide) || (isSlotSide && c.ForcePlugSide) {
```

Comment:
Indeed this can be simplified a bit, but this suggestion goes too far, it breaks the handling of "--slot" on the plug side. I've changed this to:
if c.ForcePlugSide || (isPlugSide && !c.ForceSlotSide)